aiida/engine/processes/workchains/utils.py

```diff
@@ -0,0 +1,118 @@
+# -*- coding: utf-8 -*-
+"""Utilities for `WorkChain` implementations."""
+from collections import namedtuple
+from functools import wraps
+
+from ..exit_code import ExitCode
+
+__all__ = ('ProcessHandler', 'ProcessHandlerReport', 'register_process_handler')
+
+ProcessHandler = namedtuple('ProcessHandler', 'method priority exit_codes')
+ProcessHandler.__new__.__defaults__ = (None, 0, None)
+"""A namedtuple to define a process handler for a :class:`aiida.engine.BaseRestartWorkChain`.
+
+The priority determines in which order the handlers methods are executed, with the higher priority being executed
+first. The method defines an unbound `WorkChain` method that takes an instance of a :class:`~aiida.orm.ProcessNode` as
+its sole argument. The method can return an optional `ProcessHandlerReport` to signal whether other handlers still need
+to be considered or whether the workchain should be terminated immediately.
+
+:param method: the workchain class method
+:param priority: integer denoting the process handler's priority
+:param exit_codes: single or list of `ExitCode` instances. A handler that defines this should only be called for a given
+    completed process if its exit status is a member of `exit_codes`.
+"""
+
+ProcessHandlerReport = namedtuple('ProcessHandlerReport', 'do_break exit_code')
+ProcessHandlerReport.__new__.__defaults__ = (False, ExitCode())
+"""A namedtuple to define a process handler report for a :class:`aiida.engine.BaseRestartWorkChain`.
+
+This namedtuple should be returned by a process handler of a workchain instance if the condition of the handler was
+met by the completed process. If no further handling should be performed after this method the 'do_break' field should
+be set to `True`. If the handler encountered a fatal error and the work chain needs to be terminated, an `ExitCode` with
+non-zero exit status can be set. This exit code is what will be set on the work chain itself.
+
+:param do_break: boolean, set to `True` if no further process handlers should be called, default is `False`
+:param exit_code: an instance of the :class:`~aiida.engine.processes.exit_code.ExitCode` tuple
```

Comment:
Maybe say that by default the `ExitCode` is 0 which means succesfull error handling.